src/radical/pilot/agent/agent_0.py

```diff
@@ -312,129 +322,124 @@ def finalize(self):
-                #        not provided, then we don't wait - this will be
-                #        replaced with another callback (BaseComponent.advance will
-                #        publish control command "service_up" for service tasks)
-                # FIXME: wait at least for AGENT_EXECUTING state
-
-            elif os.path.isfile(startup_file):
-                # if file exists then service is up (general approach)
-                service_up = True
-
-                # collect data from the startup file: at this point we look
-                # for URLs only
-                service_urls = {}
-                with ru.ru_open(startup_file, 'r') as fin:
-                    for line in fin.readlines():
-                        if '://' not in line:
-                            continue
-                        parts = line.split()
-                        if len(parts) == 1:
-                            idx, url = '', parts[0]
-                        elif '://' in parts[1]:
-                            idx, url = parts[0], parts[1]
-                        else:
-                            continue
-                        service_urls[idx] = url
-
-                if service_urls:
-                    for idx, url in service_urls.items():
-                        key = cb_data[tid]['name']
-                        if idx:
-                            key += '.%s' % idx
-                        key += '.url'
-                        self.session._reg[key] = url
-
-            if service_up:
-                self.publish(rpc.CONTROL_PUBSUB, {'cmd': 'service_up',
-                                                  'arg': {'uid': tid}})
-                del cb_data[tid]
+        with self._service_lock:
+
+            self._service_start_evt.clear()
+
+            self._log.debug('set agent service id to %s', tid)
+            self._service_uid_launched = tid
+            self._service_infos[tid] = dict()
+
+            self.advance(task, publish=False, push=True)
+
+            # At this point we wait for one of two events to happen: either the
+            # task will go into a state beyond `AGENT_EXECUTING` and will thus
+            # have completed prematurely in which case we declare a failure, or
+            # the service comes up and the wrapper sends a control message.  In
+            # the former case, the state cb will trigger the service startup
+            # event, in the latter case, that message will trigger that event as
+            # well.
+            #
+            # Or we time out of course :-)
+            #
+            # FIXME: need to watch state updates
+
+            # The service task timeout is watched by the wrapper, but we also
+            # watch the wrapper itself.
+            if orig_timeout:
+                if not self._service_start_evt.wait(timeout=orig_timeout):
+                    raise RuntimeError('Unable to start service')
+            else:
+                self._service_start_evt.wait()
```

Comment:
should this be skipped? assuming if timeout is not set then consider it is up right away

Reply:
hmmm, seems that `self._service_start_evt` is set only after service is done (!)

p.s. actually one service task was running, while the second one started only at the end, which confirms the point above

Reply:
I am not sure I understand that point.  The event is set in [line 686](https://github.com/radical-cybertools/radical.pilot/blob/52ed350d93cf5762ba04c4e83562f3f7a67f611e/src/radical/pilot/agent/agent_0.py#L686) when the respective control message is received.  That message in turn is sent by the control wrapper when the respective service info was obtained [here](https://github.com/radical-cybertools/radical.pilot/blob/52ed350d93cf5762ba04c4e83562f3f7a67f611e/bin/radical-pilot-service-wrapper#L231).  What test did you run, and do you have any idea why the above would not have happened?

Reply:
ah, looks like in my experiment it was waiting for a pattern
```
1729205126.74 watching: service.0000.stdout.0000 for pattern [\<http[s]?://[0-9a-zA-Z\.:_-]*\>]
1729205126.74 watching: <_io.TextIOWrapper name='service.0000.stdout.0000' mode='r' encoding='UTF-8'>
```
but there was no pattern, and service-wrapper used a default one
https://github.com/radical-cybertools/radical.pilot/blob/1a184287c23fbda9f9b102fcefa734321cbf58f7/bin/radical-pilot-service-wrapper#L75
while this to be fixed, I still would propose to have some default timeout for `self._service_start_evt.wait()`. 
Also, maybe to have a dedicated TD attribute `service_info` which represents service properties (startup-,exec-related), which will include `pattern` and `startup_timeout`. (Having different meanings for an attribute would be confusing.)

Reply:
The attribute is now named `startup_timeout`.